src/parquet/arrow/schema.cc

```diff
@@ -285,11 +343,22 @@ Status FromParquetSchema(const SchemaDescriptor* parquet_schema,
   // from the root Parquet node
   const GroupNode* schema_node = parquet_schema->group_node();
 
-  int num_fields = static_cast<int>(column_indices.size());
+  // Put the right leaf nodes in an unordered set
+  int num_columns = static_cast<int>(column_indices.size());
+  std::unordered_set<NodePtr> included_leaf_nodes(num_columns);
+  for (int i = 0; i < num_columns; i++) {
+    auto column_desc = parquet_schema->Column(column_indices[i]);
+    included_leaf_nodes.insert(column_desc->schema_node());
+  }
 
-  std::vector<std::shared_ptr<Field>> fields(num_fields);
-  for (int i = 0; i < num_fields; i++) {
-    RETURN_NOT_OK(NodeToField(schema_node->field(column_indices[i]), &fields[i]));
+  std::vector<std::shared_ptr<Field>> fields;
+  std::shared_ptr<Field> field;
+  for (int i = 0; i < schema_node->field_count(); i++) {
```

Comment:
@advancedxy Sorry I'm working on too many things right now... would be very happy too if you send a PR
(@wesm)